packages/docgen/lib/get-type-annotation.js

```diff
@@ -0,0 +1,424 @@
+		}
+		case 'TSTypeQuery': {
+			// unsure what this is
+			return '';
+		}
+		case 'TSTypeReference': {
+			return getTypeReferenceTypeAnnotation( typeAnnotation );
+		}
+		case 'TSUndefinedKeyword': {
+			return 'undefined';
+		}
+		case 'TSUnionType': {
+			return getUnionTypeAnnotation( typeAnnotation );
+		}
+		case 'TSUnknownKeyword': {
+			return 'unknown';
+		}
+		case 'TSVoidKeyword': {
+			return 'void';
+		}
+		default: {
+			return '';
+		}
+	}
+}
+
+/**
+ * @param {CommentTag} tag The documented parameter.
+ * @param {ASTNode} token The function the parameter is documented on.
+ * @return {null | string} The parameter's type annotation.
+ */
+function getParamTypeAnnotation( tag, token ) {
+	// search for the param token corresponding to the tag
+	// use the tag's `name` property
+
+	if ( babelTypes.isExportNamedDeclaration( token ) ) {
+		token = token.declaration;
+	}
+
+	if ( babelTypes.isVariableDeclaration( token ) ) {
+		// ignore multiple variable declarations
+		token = token.declarations[ 0 ].init;
+	}
+
+	// otherwise find the corresponding parameter token for the documented parameter
+	/** @type {babelTypes.Identifier} */
+	const paramToken = token.params.reduce( ( found, pToken ) => {
+		if ( found ) return found;
+		return pToken.name === tag.name ? pToken : found;
+	}, null );
+
+	// This shouldn't happen due to our ESLint enforcing correctly documented parameter names but just in case
+	// we'll give a descriptive error so that it's easy to diagnose the issue.
+	if ( ! paramToken ) {
+		throw new Error(
+			`Cound not find corresponding parameter token for documented parameter ${ tag.name } in function ${ token.id.name }.`
+		);
+	}
+
+	/** @type {babelTypes.TSTypeAnnotation} */
+	const typeAnnotation = paramToken.typeAnnotation.typeAnnotation;
+
+	return getTypeAnnotation( typeAnnotation );
+}
+
+/**
+ * @param {CommentTag} tag A return comment tag.
+ * @param {ASTNode} token A function token.
+ * @return {null | string} The function's return type annoation.
+ */
+function getReturnTypeAnnotation( tag, token ) {
+	if ( babelTypes.isExportNamedDeclaration( token ) ) {
+		token = token.declaration;
+	}
+
+	if ( babelTypes.isVariableDeclaration( token ) ) {
+		// ignore multiple variable declarations
+		token = token.declarations[ 0 ].init;
+	}
+
+	return getTypeAnnotation( token.returnType.typeAnnotation );
+}
+
+module.exports =
+	/**
+	 * @param {CommentTag} tag A comment tag.
+	 * @param {ASTNode} token A function token.
+	 * @return {null | string} The type annotation for the given tag or null if the tag has no type annotation.
+	 */
+	function ( tag, token ) {
+		// If the file is using JSDoc type annotations, use the JSDoc.
+		if ( tag.type ) {
+			return tag.type;
+		}
+
+		switch ( tag.tag ) {
+			case 'param': {
+				return getParamTypeAnnotation( tag, token );
+			}
+			case 'return': {
+				return getReturnTypeAnnotation( tag, token );
+			}
+			default: {
+				return tag.type;
```

Comment:
I lifted the `tag.type` check here and that uncovered this path: this is when a type wasn't documented by JSDoc and there's no TS info we can extract from the ASTNode either. I suppose return an empty string or "unknown type" should be fine here.

Reply:
Perhaps empty is a better choice to follow the conventions at https://github.com/WordPress/gutenberg/pull/29189/files?file-filters%5B%5D=.js&file-filters%5B%5D=.md&file-filters%5B%5D=.tsx&hide-deleted-files=true#diff-5c6936b9e1276b628a49212453264bc8f663f1f964dea9badeefb73216f4a8c5R338

Reply:
That's a good point. I think in that case an empty string is fine.